transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapChannel.java

```diff
@@ -0,0 +1,300 @@
+                in.remove(e);
+            }
+        }
+    }
+
+    @Override
+    protected AbstractEpollUnsafe newUnsafe() {
+        return new TunTapChannelUnsafe();
+    }
+
+    final LinuxSocket tunTapFd() {
+        return (LinuxSocket) fd();
+    }
+
+    final class TunTapChannelUnsafe extends AbstractEpollUnsafe {
+
+        private final List<TunTapPacket> rcvdPackets = new ArrayList<TunTapPacket>();
+
+        @Override
+        public void connect(SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise) {
+            throw new UnsupportedOperationException("TunTapChannelUnsafe.connect");
+        }
+
+        @Override
+        void epollInReady() {
+            assert eventLoop().inEventLoop();
+            if (tunTapFd().isInputShutdown()) {
+                clearEpollIn0();
+                return;
+            }
+
+            EpollRecvByteAllocatorHandle allocHandle = recvBufAllocHandle();
+            allocHandle.edgeTriggered(isFlagSet(Native.EPOLLET));
+
+            ByteBufAllocator allocator = config.getAllocator();
+
+            // Tell the allocator handle we're about to start another read loop.
+            allocHandle.reset(config);
+
+            epollInBefore();
+
+            try {
+                Throwable exception = null;
+                ByteBuf packetData = null;
+                LinuxSocket tunTapFd = tunTapFd();
+
+                try {
+                    // Repeatedly read packets from the interface for as long as there are
+                    // packets to be read and the allocator handle tells us its okay...
+                    do {
+                        // Allocate an appropriately sized buffer in which to receive data.
+                        packetData = allocHandle.allocate(allocator);
+
+                        // Tell the allocator handle how much data we're attempting to read.
+                        allocHandle.attemptedBytesRead(packetData.writableBytes());
+
+                        int writerIndex = packetData.writerIndex();
+                        long readRes;
+
+                        // If the buffer has an accessible memory address, read packet data directly into the
+                        // buffer memory. Otherwise, get the internal NIO buffer associated with the ByteBuf
+                        // and read into that.
+                        if (packetData.hasMemoryAddress()) {
+                            readRes = tunTapFd.readTunTapPacketAddress(packetData.memoryAddress(), writerIndex,
+                                packetData.capacity());
+                        } else {
+                            ByteBuffer nioData = packetData.internalNioBuffer(writerIndex, packetData.writableBytes());
+                            readRes = tunTapFd.readTunTapPacket(nioData, nioData.position(), nioData.limit());
+                        }
+
+                        // Stop reading if no packet was received.
+                        if (readRes == 0) {
```

Comment:
consider making the contents/structure of this loop consistent with `AbstractEpollStreamChannel#epollInReady()` (minus the splice handling) and `AbstractKqueueStreamChannel#readReady(..)` if possible